### PR TITLE
pgwire: auto-upgrade password authentication with SCRAM hashes  (scram 9/10)

### PR DESCRIPTION
**Pull request by knz**
(PR peeled away from #74301; previous PR in series #74848; next PR in series: #74850)
Epic CRDB-5349

Release note (security update): When using the HBA authentication
method `cert-password` (the default) for SQL client connections, and
the SQL client does not present a TLS client certificate to the
server, CockroachDB now automatically upgrades the password handshake
protocol to use SCRAM-SHA-256 if the user's stored password uses the
SCRAM encoding.

The previous behavior of requesting a cleartext password is still used
if the stored password is encoded using the CRDB-BCRYPT format.

An operator can force to _always_ request SCRAM-SHA-256 when a TLS
client cert is not provided (so as to guarantee the various
security benefits of SCRAM) using the authentication methods
`cert-scram-sha-256` (either TLS client cert _or_ SCRAM-SHA-256); and
`scram-sha-256` (only SCRAM-SHA-256).

(As previously, mandatory cleartext password authentication
can be requested, e.g. for debugging purposes, by using the HBA
method `password`.)

This automatic protocol upgrade can be manually disabled
using the new cluster setting
`server.user_login.cert_password_method.auto_scram_promotion.enable` (by
setting it to `false`), for example if certain client drivers
are found to not support SCRAM-SHA-256 authentication properly.

Release note (sql change): When possible, CockroachDB will now
automatically require the PostgreSQL-compatible SCRAM-SHA-256 protocol
when performing password validation when SQL client log in. (Reminder:
this mechanism is not used when SQL clients use TLS client certs,
which is the recommended approach.)

This assumes support for SCRAM-SHA-256 in client drivers.
As of 2020, we have found this to be prevalent in the pg driver
ecosystem. However, users should be mindful of the following possible
behavior changes:

- an application that tries to detect whether password verification
  has failed by checking server error messages, might observe
  different error messages with SCRAM-SHA-256. Those checks, if
  present, need to be updated.

- if a client driver simply does not support SCRAM-SHA-256 at all,
  the operator retains the option to set the cluster setting
  `server.user_login.cert_password_method.auto_scram_promotion.enable`
  to `false` to force the previous password verification method instead.

### CODE REVIEW

**Review by rafiss — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74849)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @bdarnell and @knz)

___
*[pkg/sql/pgwire/auth_methods.go, line 233 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/74849#-MtcsA028Hsr_oY8KaJL:-MtcsA028Hsr_oY8KaJM:b37pnhu) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d4ff5764ffee1c03671348d606bd670d7ed396e/pkg/sql/pgwire/auth_methods.go#L233)):*
> ```Go
> }
> 
> // passwordAuthenticator is the authenticator function for the
> ```

nit: passwordAuthenticator -> scramAuthenticator
___
*[pkg/sql/pgwire/auth_methods.go, line 494 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/74849#-MtcrQ2x-nOhMZLHtqk5:-MtcrQ2x-nOhMZLHtqk6:b-1iqp7d) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d4ff5764ffee1c03671348d606bd670d7ed396e/pkg/sql/pgwire/auth_methods.go#L494)):*
> ```Go
> 		// The authenticator function needs a closure to retrieve the same information.
> 		// Memoize it.
> 		newpwfn := func(ctx context.Context) (bool, security.PasswordHash, error) { return expired, hashedPassword, err }
> ```

so just to confirm, both `expired` and `err` are checked later when `newpwfn` is called? perhaps a comment here that reiterates that would help
___
*[pkg/sql/pgwire/auth_methods.go, line 499 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/74849#-Mtctm8ADHyqh7v8uU9s:-Mtctm8I52TfSxFEgpuP:b-tdipaw) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d4ff5764ffee1c03671348d606bd670d7ed396e/pkg/sql/pgwire/auth_methods.go#L499)):*
> ```Go
> 		if err == nil && hashedPassword.Method() == security.HashSCRAMSHA256 {
> 			// Yes: use a SCRAM-SHA-256 exchange.
> 			c.LogAuthInfof(ctx, "no client cert but SCRAM credentials found, proceeding with SCRAM-SHA-256")
> ```

i wonder if the logging here assumes too much. currently this is only called in the `cert-password` mode, but if the code is refactored and this gets used elsewhere, the "no client cert" part of the message might be confusing.

in other words, can we have `c.LogAuthInfof(no client cert ...)` in the `authCertPassword` function, and keep `c.LogAuthInfof(proceeding with SCRAM ...)` in this function? (ditto for `c.LogAuthInfof(requesting cleartext ...)`

**Review by knz — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74849)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @bdarnell and @rafiss)

___
*[pkg/sql/pgwire/auth_methods.go, line 233 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/74849#-MtcsA028Hsr_oY8KaJL:-MtmaApMC2usf8ZS3E0g:bt2vr1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d4ff5764ffee1c03671348d606bd670d7ed396e/pkg/sql/pgwire/auth_methods.go#L233)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: passwordAuthenticator -> scramAuthenticator
</blockquote></details>

Fixed.
___
*[pkg/sql/pgwire/auth_methods.go, line 494 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/74849#-MtcrQ2x-nOhMZLHtqk5:-MtmafvaD2wjDZxKBHtQ:bombmv2) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d4ff5764ffee1c03671348d606bd670d7ed396e/pkg/sql/pgwire/auth_methods.go#L494)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

so just to confirm, both `expired` and `err` are checked later when `newpwfn` is called? perhaps a comment here that reiterates that would help
</blockquote></details>

Good call. Done.
___
*[pkg/sql/pgwire/auth_methods.go, line 499 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/74849#-Mtctm8ADHyqh7v8uU9s:-MtmbVfA4VgN4eKa2yPn:b-fflhqt) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d4ff5764ffee1c03671348d606bd670d7ed396e/pkg/sql/pgwire/auth_methods.go#L499)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i wonder if the logging here assumes too much. currently this is only called in the `cert-password` mode, but if the code is refactored and this gets used elsewhere, the "no client cert" part of the message might be confusing.

in other words, can we have `c.LogAuthInfof(no client cert ...)` in the `authCertPassword` function, and keep `c.LogAuthInfof(proceeding with SCRAM ...)` in this function? (ditto for `c.LogAuthInfof(requesting cleartext ...)`
</blockquote></details>

Great idea. Done.

**Review by rafiss — APPROVED** [review]
lgtm!

> Reviewed 8 of 8 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74849#-:-MtnaQ5w5euAiUks_6ug:b-7nbn20)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @bdarnell and @rafiss)

**Comment by knz** [comment]
NB: I'm holding off merging this PR until all the PRs in the sequence are ready (otherwise, the entire chain will be auto-closed by github).
